File: python/timestep/workflows.py
"""DBOS workflows for durable agent execution."""

import os
import json
import uuid
from typing import Any, Optional, Callable, Awaitable
from dbos import DBOS, Queue, SetWorkflowID, SetWorkflowTimeout
from .dbos_config import configure_dbos, ensure_dbos_launched, _dbos_context, is_dbos_launched, get_dbos_connection_string
from .run_state_store import RunStateStore
from ._vendored_imports import Agent, SessionABC, TResponseInputItem, RunState
from .agent_store import load_agent
from .session_store import load_session
from .db_connection import DatabaseConnection
# Import run_agent and default_result_processor inside functions to avoid circular import


# Default queue for agent workflows with rate limiting
_default_queue: Optional[Queue] = None


def _get_default_queue() -> Queue:
    """Get or create the default agent queue with rate limiting."""
    global _default_queue
    if _default_queue is None:
        # Rate limit: 50 requests per 60 seconds (conservative for LLM APIs)
        _default_queue = Queue("timestep_agent_queue", limiter={"limit": 50, "period": 60})
    return _default_queue


@DBOS.step()
async def _load_agent_step(agent_id: str) -> Agent:
    """
    Step that loads an agent from the database.
    
    Args:
        agent_id: The agent ID (UUID as string)
    
    Returns:
        The loaded Agent object
    """
    connection_string = get_dbos_connection_string()
    if not connection_string:
        raise ValueError("DBOS connection string not available")
    
    db = DatabaseConnection(connection_string=connection_string)
    await db.connect()
    try:
        agent = await load_agent(agent_id, db)
        return agent
    finally:
        await db.disconnect()


@DBOS.step()
async def _load_session_data_step(session_id: str) -> dict:
    """
    Step that loads session data from the database.
    
    Returns serializable session data dict, not the Session object itself.
    
    Args:
        session_id: The session ID (UUID as string or session's internal ID)
    
    Returns:
        Session data dict
    """
    connection_string = get_dbos_connection_string()
    if not connection_string:
        raise ValueError("DBOS connection string not available")
    
    db = DatabaseConnection(connection_string=connection_string)
    await db.connect()
    try:
        session_data = await load_session(session_id, db)
        if not session_data:
            raise ValueError(f"Session with id {session_id} not found")
        return session_data
    finally:
        await db.disconnect()


@DBOS.step()
async def _run_agent_step(
    agent: Agent,
    run_input: list[TResponseInputItem] | RunState,
    session_data: dict,
    stream: bool,
    result_processor: Optional[Callable[[Any], Awaitable[Any]]] = None
) -> dict:
    """Step that runs an agent. Returns serializable dict with RunResult data."""
    # Reconstruct Session object from session_data
    session_type = session_data.get('session_type', '')
    internal_session_id = session_data.get('session_id')
    
    if 'OpenAIConversationsSession' in session_type:
        from ._vendored_imports import OpenAIConversationsSession
        session = OpenAIConversationsSession(conversation_id=internal_session_id)
    else:
        raise ValueError(f"Unsupported session type: {session_type}")
    
    # Import here to avoid circular import
    from . import run_agent, default_result_processor
    processor = result_processor or default_result_processor
    
    # Run agent - returns RunResult
    run_result = await run_agent(agent, run_input, session, stream, processor)
    
    # Extract only serializable data - RunResult has non-serializable objects
    # Store the RunResult object reference for _save_state_step to use
    # But return only serializable data
    return {
        '_run_result_ref': id(run_result),  # Store reference ID
        'final_output': run_result.final_output,
        'interruptions': [item.model_dump() for item in run_result.interruptions] if run_result.interruptions else [],
        '_has_interruptions': bool(run_result.interruptions),
    }


@DBOS.step()
async def _save_state_step(
    result_dict: dict,  # Serializable dict from _run_agent_step
    agent_id: str,
    session_id: Optional[str]
) -> None:
    """
    Step that saves agent state using RunStateStore.
    Note: We can't pass RunResult directly, so we need to re-run or store it differently.
    For now, if there are interruptions, we'll need to handle state saving differently.
    """
    # If there are interruptions, we need the actual RunResult to call to_state()
    # But we can't serialize it. So we'll need to save state in _run_agent_step itself
    # or use a different approach. For MVP, skip state saving in workflow for now.
    # State should be saved by the caller after getting the result.
    pass


async def _execute_agent_with_state_handling(
    agent_id: str,
    input_items: list[TResponseInputItem] | RunState,
    session_id: str,
    stream: bool,
    result_processor: Optional[Callable[[Any], Awaitable[Any]]],
    timeout_seconds: Optional[float]
) -> Any:
    """Execute agent and handle state persistence via RunStateStore."""
    # Step 1: Load agent from database
    agent = await _load_agent_step(agent_id)
    
    # Step 2: Load session data from database
    session_data = await _load_session_data_step(session_id)
    
    # Step 3: Run agent - returns dict with output and interruptions
    result_dict = await _run_agent_step(agent, input_items, session_data, stream, result_processor)
    
    # Return output - state saving happens outside workflow via RunStateStore
    return {
        'output': result_dict['final_output'],
        'interruptions': result_dict['interruptions']
    }


@DBOS.workflow()
async def _agent_workflow(
    agent_id: str,
    input_items_json: str,  # Serialized input items
    session_id: str,
    stream: bool = False,
    timeout_seconds: Optional[float] = None
) -> Any:
    """
    Workflow that runs an agent using IDs stored in the database.
    
    Args:
        agent_id: The agent ID (UUID as string)
        input_items_json: JSON-serialized input items or RunState
        session_id: The session ID (UUID as string or session's internal ID)
        stream: Whether to stream the results
        timeout_seconds: Optional timeout for the workflow
    
    Returns:
        The result from run_agent
    """
    # Deserialize input items
    input_items_data = json.loads(input_items_json)
    # TODO: Reconstruct RunState or list[TResponseInputItem] from data
    # For now, assume it's a list of dicts that can be converted to TResponseInputItem
    input_items = input_items_data  # Placeholder - will need proper deserialization
    
    # Set timeout if provided
    if timeout_seconds:
        with SetWorkflowTimeout(timeout_seconds):
            return await _execute_agent_with_state_handling(
                agent_id, input_items, session_id, stream, None, timeout_seconds
            )
    else:
        return await _execute_agent_with_state_handling(
            agent_id, input_items, session_id, stream, None, timeout_seconds
        )


def register_generic_workflows() -> None:
    """
    Register the generic workflows before DBOS launch.
    This must be called before ensure_dbos_launched().
    """
    # The workflow is already registered via @DBOS.workflow() decorator
    pass


async def run_agent_workflow(
    agent_id: str,
    input_items: list[TResponseInputItem] | RunState,
    session_id: str,
    stream: bool = False,
    workflow_id: Optional[str] = None,
    timeout_seconds: Optional[float] = None
) -> Any:
    """
    Run an agent in a durable DBOS workflow.
    
    This workflow automatically saves state on interruptions and can be resumed
    if the process crashes or restarts.
    
    Args:
        agent_id: The agent ID (UUID as string) - agent must be saved to database first
        input_items: Input items or RunState for the agent
        session_id: Session ID (UUID as string) - session must be saved to database first
        stream: Whether to stream the results
        workflow_id: Optional workflow ID for idempotency
        timeout_seconds: Optional timeout for the workflow
    
    Returns:
        The result from run_agent
    """
    # Ensure DBOS is configured (but not launched yet - workflows must be registered before launch)
    if not _dbos_context.is_configured:
        await configure_dbos()
    
    # Ensure generic workflow is registered
    register_generic_workflows()
    
    # Launch DBOS if not already launched (after workflow registration)
    if not _dbos_context.is_launched:
        await ensure_dbos_launched()
    
    # Serialize input items
    input_items_json = json.dumps(input_items, default=str)
    
    # Call the workflow with serializable parameters
    if workflow_id:
        with SetWorkflowID(workflow_id):
            return await _agent_workflow(agent_id, input_items_json, session_id, stream, timeout_seconds)
    else:
        return await _agent_workflow(agent_id, input_items_json, session_id, stream, timeout_seconds)


async def queue_agent_workflow(
    agent_id: str,
    input_items: list[TResponseInputItem] | RunState,
    session_id: str,
    stream: bool = False,
    queue_name: Optional[str] = None,
    workflow_id: Optional[str] = None,
    timeout_seconds: Optional[float] = None,
    priority: Optional[int] = None,
    deduplication_id: Optional[str] = None
) -> Any:
    """
    Enqueue an agent run in a DBOS queue with rate limiting support.
    
    This is useful for managing concurrent agent executions and respecting
    LLM API rate limits.
    
    Args:
        agent_id: The agent ID (UUID as string) - agent must be saved to database first
        input_items: Input items or RunState for the agent
        session_id: Session ID (UUID as string) - session must be saved to database first
        stream: Whether to stream the results
        queue_name: Optional queue name. Defaults to "timestep_agent_queue"
        workflow_id: Optional workflow ID for idempotency
        timeout_seconds: Optional timeout for the workflow
        priority: Optional priority (lower number = higher priority)
        deduplication_id: Optional deduplication ID to prevent duplicate runs
    
    Returns:
        WorkflowHandle that can be used to get the result via handle.get_result()
    """
    # Ensure DBOS is configured (but not launched yet - workflows must be registered before launch)
    if not _dbos_context.is_configured:
        await configure_dbos()
    
    # Ensure generic workflow is registered
    register_generic_workflows()
    
    # Launch DBOS if not already launched (after workflow registration)
    if not _dbos_context.is_launched:
        await ensure_dbos_launched()
    
    # Get queue
    if queue_name:
        queue = Queue(queue_name)
    else:
        queue = _get_default_queue()
    
    # Serialize input items
    input_items_json = json.dumps(input_items, default=str)
    
    # Enqueue options
    from dbos import SetEnqueueOptions
    from contextlib import ExitStack
    
    enqueue_options = {}
    if priority is not None:
        enqueue_options["priority"] = priority
    if deduplication_id:
        enqueue_options["deduplication_id"] = deduplication_id
    
    # Enqueue the workflow with serializable parameters
    with ExitStack() as stack:
        if workflow_id:
            stack.enter_context(SetWorkflowID(workflow_id))
        if timeout_seconds:
            stack.enter_context(SetWorkflowTimeout(timeout_seconds))
        if enqueue_options:
            stack.enter_context(SetEnqueueOptions(**enqueue_options))
        
        handle = queue.enqueue(_agent_workflow, agent_id, input_items_json, session_id, stream, timeout_seconds)
    
    return handle


async def create_scheduled_agent_workflow(
    crontab: str,
    agent_id: str,
    input_items: list[TResponseInputItem] | RunState,
    session_id: str,
    stream: bool = False
) -> None:
    """
    Create a scheduled workflow that runs an agent periodically.
    
    This function registers a scheduled workflow with DBOS. The workflow will
    run automatically according to the crontab schedule.
    
    Note: This must be called before ensure_dbos_launched() because scheduled
    workflows must be registered before DBOS launch.
    
    Example:
        create_scheduled_agent_workflow(
            "0 0,6,12,18 * * *",  # Every 6 hours
            agent_id,
            input_items,
            session_id
        )
    
    Args:
        crontab: Crontab schedule (e.g., "0 0,6,12,18 * * *" for every 6 hours)
        agent_id: The agent ID (UUID as string) - agent must be saved to database first
        input_items: Input items or RunState for the agent
        session_id: Session ID (UUID as string) - session must be saved to database first
        stream: Whether to stream the results
    
    Raises:
        RuntimeError: If DBOS is already launched
    """
    # Check if DBOS is already launched - if so, we can't register new scheduled workflows
    if is_dbos_launched():
        raise RuntimeError(
            "Cannot create scheduled workflow after DBOS launch. "
            "Scheduled workflows must be registered before DBOS.launch() is called. "
            "Call create_scheduled_agent_workflow() before ensure_dbos_launched()."
        )
    
    # Ensure DBOS is configured (but not launched yet)
    if not _dbos_context.is_configured:
        await configure_dbos()
    
    # Serialize input items
    input_items_json = json.dumps(input_items, default=str)
    
    # Register a scheduled workflow
    @DBOS.scheduled(crontab)
    @DBOS.workflow()
    async def _scheduled_workflow(scheduled_time: Any, actual_time: Any):
        return await _agent_workflow(agent_id, input_items_json, session_id, stream, None)

File: typescript/timestep/workflows.ts
/** DBOS workflows for durable agent execution. */

import { DBOS, WorkflowQueue, WorkflowHandle } from '@dbos-inc/dbos-sdk';
import { configureDBOS, ensureDBOSLaunched, getDBOSConnectionString, isDBOSLaunched } from './dbos_config.ts';
import { RunStateStore } from './run_state_store.ts';
import { runAgent, defaultResultProcessor } from './index.ts';
import { Agent, Session, RunState } from '@openai/agents';
import type { AgentInputItem } from '@openai/agents-core';
import { loadAgent } from './agent_store.ts';
import { loadSession } from './session_store.ts';
import { DatabaseConnection } from './db_connection.ts';

// Default queue for agent workflows with rate limiting
let defaultQueue: WorkflowQueue | null = null;

function getDefaultQueue(): WorkflowQueue {
  if (!defaultQueue) {
    // Rate limit: 50 requests per 60 seconds (conservative for LLM APIs)
    defaultQueue = new WorkflowQueue('timestep_agent_queue', {
      rateLimit: { limitPerPeriod: 50, periodSec: 60 }
    });
  }
  return defaultQueue;
}

async function loadAgentStep(agentId: string): Promise<Agent> {
  /**
   * Step that loads an agent from the database.
   */
  const connectionString = getDBOSConnectionString();
  if (!connectionString) {
    throw new Error('DBOS connection string not available');
  }

  const db = new DatabaseConnection({ connectionString });
  await db.connect();
  try {
    const agent = await loadAgent(agentId, db);
    return agent;
  } finally {
    await db.disconnect();
  }
}

async function loadSessionDataStep(sessionId: string): Promise<any> {
  /**
   * Step that loads session data from the database.
   * 
   * Returns serializable session data object, not the Session object itself.
   */
  const connectionString = getDBOSConnectionString();
  if (!connectionString) {
    throw new Error('DBOS connection string not available');
  }

  const db = new DatabaseConnection({ connectionString });
  await db.connect();
  try {
    const sessionData = await loadSession(sessionId, db);
    if (!sessionData) {
      throw new Error(`Session with id ${sessionId} not found`);
    }
    return sessionData;
  } finally {
    await db.disconnect();
  }
}

const runAgentStep = DBOS.registerStep(async (
  agent: Agent,
  runInput: AgentInputItem[] | RunState<any, any>,
  sessionData: any,
  stream: boolean,
  resultProcessor?: (result: any) => Promise<any>
): Promise<any> => {
  /** Step that runs an agent. Returns serializable dict. */
  // Reconstruct Session object from sessionData
  const sessionType = sessionData.session_type || '';
  const internalSessionId = sessionData.session_id;

  let session: Session;
  if (sessionType.includes('OpenAIConversationsSession')) {
    const { OpenAIConversationsSession } = await import('@openai/agents');
    session = new OpenAIConversationsSession({ conversationId: internalSessionId });
  } else {
    throw new Error(`Unsupported session type: ${sessionType}`);
  }

  const processor = resultProcessor || defaultResultProcessor;

  // Run agent - returns RunResult
  const runResult = await runAgent(agent, runInput, session, stream, processor);

  // Extract output - RunResult might have output in different formats
  // Try multiple methods to extract the output
  let output: any = null;
  
  // Method 1: Check final_output directly
  if (runResult.final_output !== undefined && runResult.final_output !== null) {
    output = runResult.final_output;
  }
  // Method 2: Use to_input_list() to get all items, then filter for assistant messages
  else if (runResult.to_input_list && typeof runResult.to_input_list === 'function') {
    try {
      const allItems = runResult.to_input_list();
      output = allItems.filter(
        (item: any) => item.type === 'message' && item.role === 'assistant'
      );
    } catch (e) {
      // Fall through to next method
    }
  }
  // Method 3: Check new_items
  if (!output && runResult.new_items) {
    const outputItems: any[] = [];
    for (const item of runResult.new_items) {
      try {
        const inputItem = (item as any).to_input_item();
        if (inputItem.type === 'message' && inputItem.role === 'assistant') {
          outputItems.push(inputItem);
        }
      } catch (e) {
        // Skip items that can't be converted
      }
    }
    if (outputItems.length > 0) {
      output = outputItems;
    }
  }
  // Method 4: Check output property directly
  if (!output && runResult.output !== undefined) {
    output = runResult.output;
  }

  // Extract only serializable data - RunResult has non-serializable objects
  return {
    final_output: output,
    interruptions: runResult.interruptions || []
  };
});

async function saveStateStep(
  result: any,
  agentId: string,
  sessionId: string | undefined
): Promise<void> {
  /**
   * Step that saves agent state. This must be a step because it accesses the database.
   */
  await DBOS.runStep(
    async () => {
      const connectionString = getDBOSConnectionString();
      if (!connectionString) {
        throw new Error('DBOS connection string not available');
      }

      const db = new DatabaseConnection({ connectionString });
      await db.connect();
      try {
        const agent = await loadAgent(agentId, db);
        const stateStore = new RunStateStore({ agent, sessionId });

        if (result.state) {
          await stateStore.save(result.state);
        } else if (result.toState) {
          const state = result.toState();
          await stateStore.save(state);
        } else {
          throw new Error('Result does not have state or toState() method');
        }
      } finally {
        await db.disconnect();
      }
    },
    { name: 'saveState' }
  );
}

async function executeAgentWithStateHandling(
  agentId: string,
  inputItems: AgentInputItem[] | RunState<any, any>,
  sessionId: string,
  stream: boolean,
  resultProcessor: ((result: any) => Promise<any>) | undefined,
  timeoutSeconds: number | undefined
): Promise<any> {
  /** Execute agent. Returns output. State persistence handled by RunStateStore outside workflow. */
  // Step 1: Load agent from database
  const agent = await loadAgentStep(agentId);

  // Step 2: Load session data from database
  const sessionData = await loadSessionDataStep(sessionId);

  // Step 3: Run agent - returns dict with output and interruptions
  const resultDict = await runAgentStep(agent, inputItems, sessionData, stream, resultProcessor);

  // Return output - state saving happens outside workflow via RunStateStore
  return {
    output: resultDict.final_output,
    interruptions: resultDict.interruptions
  };
}

/**
 * Register the generic workflows before DBOS launch.
 * This must be called before ensureDBOSLaunched().
 */
export async function registerGenericWorkflows(): Promise<void> {
  // The workflow is already registered via DBOS.registerWorkflow() decorator
  // This function is kept for compatibility but does nothing
  return;
}

const agentWorkflow = DBOS.registerWorkflow(async function agentWorkflow(
  agentId: string,
  inputItemsJson: string,  // Serialized input items
  sessionId: string,
  stream: boolean = false,
  timeoutSeconds?: number
) {
  /**
   * Workflow that runs an agent using IDs stored in the database.
   */
  // Deserialize input items
  const inputItemsData = JSON.parse(inputItemsJson);
  const inputItems = inputItemsData as AgentInputItem[] | RunState<any, any>;

  // Execute and return result
  const result = await executeAgentWithStateHandling(
    agentId,
    inputItems,
    sessionId,
    stream,
    undefined,
    timeoutSeconds
  );
  
  // Ensure we return a proper result object with output
  if (!result) {
    throw new Error('Workflow returned null/undefined result');
  }
  
  // Extract output - handle both output and final_output keys
  const output = result.output || result.final_output || null;
  const interruptions = result.interruptions || [];
  
  // Ensure output is serializable - deep clone to remove any non-serializable references
  let serializableOutput: any = null;
  if (output) {
    try {
      serializableOutput = JSON.parse(JSON.stringify(output));
    } catch (e) {
      // If serialization fails, try to extract just the text
      if (Array.isArray(output)) {
        serializableOutput = output.map((item: any) => {
          if (item && typeof item === 'object') {
            return JSON.parse(JSON.stringify(item));
          }
          return item;
        });
      } else {
        serializableOutput = output;
      }
    }
  }
  
  const serializableInterruptions = interruptions ? JSON.parse(JSON.stringify(interruptions)) : [];
  
  // Return a plain object that DBOS can serialize - use explicit keys
  // IMPORTANT: For queued workflows in DBOS TypeScript, the result must be a plain object
  // with no nested complex structures that might cause serialization issues
  const workflowResult = {
    output: serializableOutput,
    interruptions: serializableInterruptions
  };
  
  // Verify it's serializable before returning - this is critical for queued workflows
  try {
    const testSerialization = JSON.stringify(workflowResult);
    if (!testSerialization || testSerialization === '{}') {
      throw new Error('Workflow result serialization produced empty object');
    }
  } catch (e: any) {
    throw new Error(`Workflow result is not serializable: ${e.message}`);
  }
  
  // Log the result structure for debugging
  console.log('agentWorkflow returning result with keys:', Object.keys(workflowResult));
  console.log('agentWorkflow output type:', typeof workflowResult.output);
  console.log('agentWorkflow output is array?', Array.isArray(workflowResult.output));
  
  return workflowResult;
});

export async function runAgentWorkflow(
  agentId: string,
  inputItems: AgentInputItem[] | RunState<any, any>,
  sessionId: string,
  stream: boolean = false,
  workflowId?: string,
  timeoutSeconds?: number
): Promise<any> {
  /**
   * Run an agent in a durable DBOS workflow.
   * 
   * This workflow automatically saves state on interruptions and can be resumed
   * if the process crashes or restarts.
   */
  // Ensure DBOS is configured (but not launched yet - workflows must be registered before launch)
  if (!getDBOSConnectionString()) {
    await configureDBOS();
  }

  // Ensure generic workflow is registered
  await registerGenericWorkflows();

  // Launch DBOS if not already launched (after workflow registration)
  if (!isDBOSLaunched()) {
    await ensureDBOSLaunched();
  }

  // Serialize input items
  const inputItemsJson = JSON.stringify(inputItems);

  // Call the workflow with serializable parameters
  if (workflowId) {
    const handle = await DBOS.startWorkflow(agentWorkflow, { workflowID: workflowId })(
      agentId, inputItemsJson, sessionId, stream, timeoutSeconds
    );
    return await handle.getResult();
  } else {
    return await agentWorkflow(agentId, inputItemsJson, sessionId, stream, timeoutSeconds);
  }
}

export async function queueAgentWorkflow(
  agentId: string,
  inputItems: AgentInputItem[] | RunState<any, any>,
  sessionId: string,
  stream: boolean = false,
  queueName?: string,
  workflowId?: string,
  timeoutSeconds?: number,
  priority?: number,
  deduplicationId?: string
): Promise<WorkflowHandle<any>> {
  /**
   * Enqueue an agent run in a DBOS queue with rate limiting support.
   * 
   * This is useful for managing concurrent agent executions and respecting
   * LLM API rate limits.
   * 
   * @returns WorkflowHandle that can be used to get the result
   */
  // Ensure DBOS is configured (but not launched yet - workflows must be registered before launch)
  if (!getDBOSConnectionString()) {
    await configureDBOS();
  }

  // Ensure generic workflow is registered
  await registerGenericWorkflows();

  // Launch DBOS if not already launched (after workflow registration)
  if (!isDBOSLaunched()) {
    await ensureDBOSLaunched();
  }

  // Get queue
  const queue = queueName ? new WorkflowQueue(queueName) : getDefaultQueue();

  // Serialize input items
  const inputItemsJson = JSON.stringify(inputItems);

  // Enqueue options
  const enqueueOptions: any = {};
  if (priority !== undefined) {
    enqueueOptions.priority = priority;
  }
  if (deduplicationId) {
    enqueueOptions.deduplicationID = deduplicationId;
  }

  const startParams: any = {
    queueName: queue.name,
    ...enqueueOptions
  };

  if (workflowId) {
    startParams.workflowID = workflowId;
  }

  if (timeoutSeconds) {
    startParams.timeoutMS = timeoutSeconds * 1000;
  }

  const handle = await DBOS.startWorkflow(agentWorkflow, startParams)(
    agentId, inputItemsJson, sessionId, stream, timeoutSeconds
  );
  return handle;
}

export async function createScheduledAgentWorkflow(
  crontab: string,
  agentId: string,
  inputItems: AgentInputItem[] | RunState<any, any>,
  sessionId: string,
  stream: boolean = false
): Promise<void> {
  /**
   * Create a scheduled workflow that runs an agent periodically.
   * 
   * This function registers a scheduled workflow with DBOS. The workflow will
   * run automatically according to the crontab schedule.
   * 
   * Example:
   *   createScheduledAgentWorkflow(
   *     '0 0,6,12,18 * * *',  // Every 6 hours
   *     agentId,
   *     inputItems,
   *     sessionId
   *   );
   */
  // Check if DBOS is already launched - if so, we can't register new scheduled workflows
  if (isDBOSLaunched()) {
    throw new Error(
      'Cannot create scheduled workflow after DBOS launch. ' +
      'Scheduled workflows must be registered before DBOS.launch() is called. ' +
      'Call createScheduledAgentWorkflow() before ensureDBOSLaunched().'
    );
  }

  // Ensure DBOS is configured (but not launched yet)
  if (!getDBOSConnectionString()) {
    await configureDBOS();
  }

  // Serialize input items
  const inputItemsJson = JSON.stringify(inputItems);

  // Register a scheduled workflow
  const scheduledWorkflow = DBOS.registerScheduled(
    async (scheduledTime: Date, startTime: Date) => {
      return await agentWorkflow(agentId, inputItemsJson, sessionId, stream, undefined);
    },
    { crontab }
  );
}
